src/tools/rbd_mirror/ImageReplayer.cc

```diff
@@ -218,26 +219,30 @@ class ImageReplayerAdminSocketHook : public AdminSocketHook {
     for (Commands::const_iterator i = commands.begin(); i != commands.end();
 	 ++i) {
       (void)admin_socket->unregister_command(i->first);
-      delete i->second;
     }
     commands.clear();
   }
 
   bool call(std::string command, cmdmap_t& cmdmap, std::string format,
 	    bufferlist& out) override {
-    Mutex::Locker locker(lock);
-    Commands::const_iterator i = commands.find(command);
-    assert(i != commands.end());
+    CommandRef command_ref;
+    {
+      Mutex::Locker locker(lock);
+      Commands::const_iterator i = commands.find(command);
+      assert(i != commands.end());
+      command_ref = i->second;
+    }
     Formatter *f = Formatter::create(format);
     stringstream ss;
-    bool r = i->second->call(f, &ss);
+    bool r = command_ref->call(f, &ss);
```

Comment:
I can use ImageReplayer lock but it does not look like it will simplify things. I will still need to acquire the lock when looking for a command and then release before calling the command (as it will acquire lock internally).

Do you want this or do you mean something different?